vhdbuilder/packer/backfill-cleanup.sh

```diff
@@ -49,28 +49,46 @@ fi
-        # clean the image-definition if ALL associated sig versions have been deleted
-        if [[ "$image_versions" == "[]" ]]; then
-          echo "Deleting sig image-definition ${image_definition} from gallery ${gallery} rg ${AZURE_RESOURCE_GROUP_NAME}"
-          az sig image-definition delete --gallery-image-definition ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME}
-        fi
-    done
-    image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
-    # clean the gallery if ALL sig image-definitions have been deleted
-    if [[ -z $image_defs ]]; then
-      echo "Deleting gallery ${gallery}"
-      az sig delete --gallery-name ${gallery} --resource-group ${AZURE_RESOURCE_GROUP_NAME}
+    if [[ "${gallery}" =~ WS2019Gallery* ]]; then
+      create_date=${gallery:13:6}
+    elif [[ "${gallery}" =~ WS2019_containerdGallery* ]]; then
+      create_date=${gallery:24:6}
+    elif [[ "${gallery}" =~ WS2022_containerdGallery* ]]; then
+      create_date=${gallery:24:6}
+    elif [[ "${gallery}" =~ WS2022_containerd_gen2Gallery* ]]; then
+      create_date=${gallery:29:6}
+    else
+      continue
+    fi
+
+    due_date=$(date +%y%m%d -d "7 days ago")
+    echo "create_date is ${create_date}"
+    echo "due_date is ${due_date}"
+    # clean the entire SIG resources if it's one week ago
+    if [[ $create_date < $due_date ]]; then
+      echo "Finding sig image definitions from gallery ${gallery}"
+      image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
+      for image_definition in $image_defs; do
+          echo "Finding sig image versions associated with ${image_definition} in gallery ${gallery}"
+          image_versions=$(az sig image-version list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} -i ${image_definition} | jq -r '.[].name')
+          for image_version in $image_versions; do
+              echo "Deleting sig image-version ${image_version} ${image_definition} from gallery ${gallery} rg ${AZURE_RESOURCE_GROUP_NAME}"
+              az sig image-version delete -e $image_version -i ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME}
+          done
+          image_versions=$(az sig image-version list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} -i ${image_definition} | jq -r '.[].name')
+          echo "image versions are $image_versions"
+          if [[ -z "${image_versions}" ]]; then
+            echo "Deleting sig image-definition ${image_definition} from gallery ${gallery} rg ${AZURE_RESOURCE_GROUP_NAME}"
+            az sig image-definition delete --gallery-image-definition ${image_definition} -r ${gallery} -g ${AZURE_RESOURCE_GROUP_NAME}
+          fi
+      done
+      image_defs=$(az sig image-definition list -g ${AZURE_RESOURCE_GROUP_NAME} -r ${gallery} | jq -r '.[] | select(.osType == "Windows").name')
```

Comment:
I think it's safe to run another check since if it's empty it won't take any extra time. But I DO observe in the past that there were some resources (sig image-def in this case) that were not deleted. And I think that's part of the reason why we need a backfill deletion because if all resources are GC as we desire we won't need this _convoluted_ logic at all at the first place.

I think run another check will make sure the definitions will not be deleted if they are being used somehow by another pipeline. And we won't need to debug the RNF (resource not found) error message.

But if running the list introduces overhead I'm happy to remove it. I just feel it's safer to double check the emptiness/existence before cleaning all.

Reply:
if for the debug purpose, I think that it may be better to log images defs if image_defs is not empty but always delete the gallery without checking it. It will fail in deleting the gallery but it has no harm, right?

Reply:
logged and deleted gallery anyway w/o checking emptiness.